src/expr/skolem_manager.cpp

```diff
@@ -50,43 +51,91 @@ Node SkolemManager::mkSkolem(Node v,
   Node predw = getWitnessForm(pred);
   // make the witness term, which should not contain any skolem
   Node w = nm->mkNode(WITNESS, bvl, predw);
-  // store the mapping to proof generator
-  d_gens[w] = pg;
+  // store the mapping to proof generator if it exists
+  if (pg != nullptr)
+  {
+    Node q = nm->mkNode(EXISTS, w[0], w[1]);
+    // notice this may overwrite an existing proof generator
+    d_gens[q] = pg;
+  }
   return getOrMakeSkolem(w, prefix, comment, flags);
 }
 
-Node SkolemManager::mkSkolemExists(Node v,
-                                   Node q,
-                                   const std::string& prefix,
-                                   const std::string& comment,
-                                   int flags,
-                                   ProofGenerator* pg)
+Node SkolemManager::mkSkolemize(Node q,
+                                std::vector<Node>& skolems,
+                                const std::string& prefix,
+                                const std::string& comment,
+                                int flags,
+                                ProofGenerator* pg)
+{
+  Trace("sk-manager-debug") << "mkSkolemize..." << std::endl;
+  Assert(q.getKind() == EXISTS);
+  Node currQ = q;
+  for (const Node& av : q[0])
+  {
+    Assert(currQ.getKind() == EXISTS && av == currQ[0][0]);
+    Node sk = skolemize(currQ, currQ, prefix, comment, flags);
```

Comment:
please add a comment to say that `currQ` will be update by `skolemize` and this will guarantee the right shape for the next step